### PR TITLE
#163764668 Test Signup and login

### PR DESCRIPTION
**Pull request by EuniceKiingati**
**What does this PR do?**
This PR writes tests for the endpoint for user login and signup

**Description of Task to be completed**
_The main tasks to be completed are:_
Creation of tests for endpoints to signup a user and login

**How should this be manually tested?**

1. Cretate  a virual enviroment with the command
   `$ virtualenv -p python3 venv`
2. Activate the venv with the command
   `$ source venv/bin/activate`
3. Install git
4. clone this repo
   `$ https://github.com/EuniceKiingati/Politico-V1"`
5. cd into the folder Politico-V1
6. export required enviroments
   `$ export FLASK_APP="run.py"`
7. install requirements
   `$ pip install -r requirements.txt`
8. now we are ready to run.
   
   1. for tests run
      `$ pytest`
   2. for the application run
      `$ flask run`

If you ran the application you can test the api end points using postman as follows
1. Signup a user by sending a `post` request to the endpoint `api/v1/users` with the signup data below
``` source-json
{
	"username":"Test",
	"email":"Test@email.com",
	"password":"Test123#"
}
```
1. Logina user by sending a `post` request to the endpoint `api/v1/users/login` with the logindata below
``` source-json
{
	"username":"Test",
	"password":"Test123#"
}
```
**What are the relevant pivotal tracker stories?**
[#163764668](https://www.pivotaltracker.com/story/show/163764668)
**Screenshots (if appropriate)**
1. User signup
screenshot 4
1. User login
image

### CODE REVIEW

File: app/app.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+                response = jsonify({
+                    "message": "User login successful",
+                    "status": 200,
+                    "data": user
+                })
+                response.status_code = 200
+                return response
+        response = jsonify({
+            "message": "User login failed, Check your credentials",
+            "status": 404,  # not  found
```
Comment (thoth-ky) [inline app/app.py]: this should return a 401 status code, UNAUTHORIZED

File: app/app.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        })
+        response.status_code = 404
+        return response
+
+    @app.route('/api/v1/parties', methods=['POST', 'GET'])
+    def create_party():
+        if request.method == 'GET':
```
Comment (thoth-ky) [inline app/app.py]: Can you split this into 2 methods instead, one for the GET and another for the POST. the url can remain the same

File: app/app.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+            for party in political_parties:
+                if party['id'] == party_id:
+                    response = jsonify({
+                        "status": 200,
+                        "data": party
+                    })
+                    response.status_code = 200
+                    return response
+            response = jsonify({
+                "message": "invalid id",
```
Comment (thoth-ky) [inline app/app.py]: Return more meaningful messages, like "party not found"

File: app/test_api.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            "password": "Yunis",
+            
+        })
+        response = self.client().post('/api/v1/users', data=UserInfo,
+                                      content_type='application/json')
+        print(response.data)
+        self.assertEqual(response.status_code, 400)
+    
+    def test_signup_string_password(self):
+        UserInfo = json.dumps({
```
Comment (thoth-ky) [inline app/test_api.py]: Observe python naming conventions, throughout
```suggestion
        user_Info = json.dumps({
```

File: app/test_api.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            "username": "sampleusername",
+            "email": "sampleemail",
+            "password": "2",
+           
+        })
+        response = self.client().post('/api/v1/users', data=UserInfo,
+                                      content_type='application/json')
+        print(response.data)
```
Comment (thoth-ky) [inline app/test_api.py]: Remove unnecessary lines of code after debugging
```suggestion
```

File: app/app.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+def bad_request(message):
+    response = jsonify({
+                "message": message,
+                "status": 400,
+                })
+    response.status_code = 400
+    return response
+
+def create_app():
```
Comment (thoth-ky) [inline app/app.py]: this file is clearly too long. Can you try split the app into different files, models, views and version the application

**Review by thoth-ky — CHANGES_REQUESTED** [review]

**Comment by EuniceKiingati** [comment]
Thanks for the feedback, let me work on it.

File: app/app.py (hunk 3 above)
Comment (EuniceKiingati) [reply app/app.py]: Fixed

File: app/test_api.py (hunk 4 above)
Comment (EuniceKiingati) [reply app/test_api.py]: Fixed

File: app/test_api.py (hunk 5 above)
Comment (EuniceKiingati) [reply app/test_api.py]: Fixed

File: app/app.py (hunk 6 above)
Comment (EuniceKiingati) [reply app/app.py]: Fixed

**Pull request merged**